research/06-require-ana.tex

```diff
@@ -1,9 +1,184 @@
+should be generated from the same definitions if it is to fit within Spoofax.
+
+\todo{The following requirement is also a design goal.}
+\paragraph{Generator should be language definition agnostic} All services
+generated by Spoofax make no assumptions about language constructs or syntax.
+This should be no different for the REPL.
+
+\paragraph{Input \& output history} Users should be able to retrieve previously
+typed expressions and statements to support the explorative and interactive
+nature of a REPL. In the same vein, previously yielded results should be
+implicitly bound to automatically generated names to make their values available
+in future expressions.
+
+\paragraph{Multiline input editing} Most, if not all, modern languages benefit
+by structuring their constructs over multine lines. Multiline input editing is
+therefore a crucial feature for user satisfaction. The input editor should
+start in single line editing mode and recognise when an expression or statement
+is part of a multiline construction. When it recognises this, it should
+automatically switch to its multiline input editing mode, wherein the
+prompt character indicates this new mode. This mode's behavior (e.g. keyboard
+shortcuts) is different from the single line editing mode.
+
+\paragraph{Error reporting} To support the interactivity of the REPL, error
+reporting should be available in two ways: while typing an expression or a
+statement, on-the-fly error reporting should indicate wrongly typed items. The
+interpreter that evaluates syntactically correct input should also be able to
+display its error messages to the user in case of an error during the dynamic
+semantics phase.
+
+\paragraph{Syntax checked input} Supporting the above requirement, all
+input should have its syntax checked on the fly.
+
+\paragraph{Syntax highlighting} All expressions and statements (whether they are
+currently being entered, displayed as previously entered input or displayed as
+previously yieled results) should be syntax highlighted.
+
+\paragraph{Integration with Eclipse} As first implementation, the generated
+REPLs should integrate with Eclipse to provide an interface to the users. It is
+important to note that while integration with Eclipse is a requirement, Spoofax
+itself is moving to be IDE-agnostic. As such, the implementation of the REPL's
+core should be generic and not be tied to Eclipse.
+
+\subsubsection{Should have}
+
+Requirements listed under ``should-haves'' are important, but not necessary for
+a working product.
+
+\paragraph{Ability to redefine identifiers} As explained in ~\cref{sec:repl},
+REPLs provide the ability to explore unknown problem domains. It is not a
+far-fetched idea that a developer would want to change function implementations
+or the types of certain variables. To support this, a REPL should allow users to
+redefine identifiers. This might mean that the REPL needs different
```

Comment:
> This might mean that the REPL needs different semantics than the language it operates on, slightly contrasting to the requirement that the REPLs should be language-agnostic.

This probably belongs in 'problem analysis'.